scio-core/src/main/scala/com/spotify/scio/coders/KryoAtomicCoder.scala

```diff
@@ -134,9 +107,36 @@ private[scio] class KryoAtomicCoder[T] extends AtomicCoder[T] {
     case (key, wrapper: JIterableWrapper[_]) =>
       observer.update(kryoEncodedElementByteSize(key))
       // FIXME: handle ElementByteSizeObservableIterable[_, _]
+      var count = 0
+      var bytes = 0L
+      var warned = false
+      var aborted = false
+      val warningThreshold = 10000 // 10s
+      val abortThreshold = 60000 // 1min
+      val start = System.currentTimeMillis()
       val i = wrapper.underlying.iterator()
-      while (i.hasNext) {
-        observer.update(kryoEncodedElementByteSize(i.next()))
+      while (i.hasNext && !aborted) {
+        val size = kryoEncodedElementByteSize(i.next())
+        observer.update(size)
+        count += 1
+        bytes += size
+        val elapsed = System.currentTimeMillis() - start
+        if (elapsed > abortThreshold) {
+          aborted = true
+          logger.warn(s"Aborting size estimation for ${wrapper.underlying.getClass}, " +
+            s"elapsed: $elapsed ms, count: $count, bytes: $bytes")
+          wrapper.underlying match {
+            case c: _root_.java.util.Collection[_] =>
+              // interpolate remaining bytes in the collection
```

Comment:
You could just in case also add a `warn` or `debug` message here for getting size of collection and how much it took, just in case it's expensive, also might be good information to have, and is only a single message anyway.